### PR TITLE
Add async HalibutRuntime.SendOutgoingHttpsRequestAsync and make ISecureClient async

### PR DESCRIPTION
**Pull request by nathanwoctopusdeploy**
# Background

Adds an async version of 

`HalibutRuntime.SendOutgoingHttpsRequest`

and 

`ISecureCLient` and all implementations of the interface.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

**Comment by nathanwoctopusdeploy** [comment]
[sc-53211]

File: source/Halibut.Tests/Transport/SecureClientFixture.cs [hunk 1]
```diff
@@ -39,8 +42,14 @@ public void Dispose()
         }
 
         [Test]
-        public void SecureClientClearsPoolWhenAllConnectionsCorrupt()
+        [SyncAndAsync]
+        public async Task SecureClientClearsPoolWhenAllConnectionsCorrupt(SyncOrAsync syncOrAsync)
         {
+            if (syncOrAsync == SyncOrAsync.Async)
+            {
+                Assert.Inconclusive("TODO - ASYNC ME UP! Code path incomplete for the system under test.");
```
Comment (LukeButters) [inline source/Halibut.Tests/Transport/SecureClientFixture.cs:50]: For my understanding when does this become async?

File: source/Halibut/HalibutRuntime.cs [hunk 2]
```diff
@@ -264,6 +265,36 @@ async Task<ResponseMessage> SendOutgoingRequestSynchronouslyAsync(RequestMessage
             return response;
         }
 
+        async Task<ResponseMessage> SendOutgoingRequestAsync(RequestMessage request, MethodInfo methodInfo, CancellationToken cancellationToken)
```
Comment (LukeButters) [inline source/Halibut/HalibutRuntime.cs:268]: I think I added this method [here](https://github.com/OctopusDeploy/Halibut/pull/356/files#diff-a5f79eb60063621324c5189327912f06a77a5ade175b6ce22430447ee41a60e4R278) I think that PR ought to get in asap.

Comment (nathanwoctopusdeploy) [reply source/Halibut/HalibutRuntime.cs:268]: Yerp, need an async entry point to get this path tested!

File: source/Halibut.Tests/Transport/SecureClientFixture.cs (hunk 1 above)
Comment (nathanwoctopusdeploy) [reply source/Halibut.Tests/Transport/SecureClientFixture.cs]: This has been changed so the test now passes. We now wire up incomplete async paths to the sync code path

File: source/Halibut.Tests/Support/TestAttributes/SyncAndAsyncAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    public enum SyncOrAsync
+    {
+        Sync,
+        Async
+    }
+
+    public static class SyncOrAsyncExtensions
+    {
+        public static SyncOrAsync WhenSync(this SyncOrAsync syncOrAsync, Action action)
```
Comment (LukeButters) [inline source/Halibut.Tests/Support/TestAttributes/SyncAndAsyncAttribute.cs:35]: Nice

**Review by LukeButters — APPROVED** [review]

**Pull request merged**